extension/commercial_data/README.md

```diff
@@ -0,0 +1,416 @@
+            enum:
+              - orderable
+              - ordered
+              - shipping
+              - delivered
+              - unable_to_deliver
+              - canceled 
+          order:date:
+            type: string
+            format: date-time
+            description: The order time. Formatted as a
+              [RFC 3339](https://www.rfc-editor.org/rfc/rfc3339.html) date-time.
+          order:expiration_date:
+            type: string
+            format: date-time
+            description: The validity time of the order. Formatted as a
+              [RFC 3339](https://www.rfc-editor.org/rfc/rfc3339.html) date-time.
+          collection_id:
+            type: string
+            description: Unique identifier of the collection.
+          costs:
+            $ref: 'https://raw.githubusercontent.com/Open-EO/openeo-api/1.0.0/openapi.yaml#/components/schemas/money'
+    links:
+      $ref: 'https://raw.githubusercontent.com/Open-EO/openeo-api/1.0.0/openapi.yaml#/components/schemas/links_pagination'
+```
+
+#### `POST /orders`
+
+Create an order for selected products. Order can contain some additional parameters that specify how the products should be delivered. For example, depending on the collection it might be possible to set the projection, resampling method, bit depth etc of the delivered data.
+
+Backends should expose the available ordering parameters in `/collections/{collection-id}` in the `order_parameters` field, following the `process_parameters` schema of [`GET /service_types`](https://docs.openeo.cloud/api/#tag/Secondary-Services/operation/list-service-types). 
+
+The request should return the `order-id`, current status and the costs of the order.
+
+```yaml
+schema:
+  title: Create Order Request
+  type: object
+  required:
+    - collection_id
+    - products
+    - parameters
+  properties:
+    collection_id:
+      type: string
+      description: Unique identifier of the collection.
+    products:
+      type: array
+      description: Array of IDs of products to order.
+      items:
+        $ref: 'https://raw.githubusercontent.com/Open-EO/openeo-api/1.0.0/openapi.yaml#/components/schemas/order_id'
+    parameters:
+      type: object
+      description: Key-value pairs of available `order_parameters` as listed at `GET /commercial_data/collections/{collection_id}` for filtering available products.
+      additionalProperties:
+        x-additionalPropertiesName: Order Parameter Name
+        description: Value of the order parameter to be used in the request.
+```
+
+#### `GET /orders/{order-id}`
+
+Get full metadata of the order. The item should follow the [STAC Order Extension](https://github.com/stac-extensions/order), but extended with the spatial and temporal extent information and other metadata about the products.
+
+```json
+{
+  "order:id": "40264b5-c3ae-46f4-a907-0f612d763d97",
+  "order:status": "delivered",
+  "order:date": "2017-01-01T09:32:12Z",
+  "products": [
```

Comment:
Instead of inlining them, we may want to consider linking to them. Also, I'm wondering whether we should generally also think about which of our additions we can propose to be added to the STAC extension?

Reply:
Thinking more about it, we really should list the products at /collections/{collection_id}/items and just link to them from here so that we don't need to define the whole STAC structure again for this endpoint.

Reply:
See also https://github.com/Open-EO/openeo-api/pull/447#discussion_r924801576

Reply:
This would be quite a problem for our implementation, as Sentinel Hub doesn't support accessing information about a specific product directly, or listing all of them for that matter. It only supports searching them (which behind the scenes uses the searching functionality of the third parties who provide the data). 

In the "previous" proposal, `GET /orders` did not include any "heavy" information  about products in order to make it lightweight: https://github.com/Open-EO/openeo-api/pull/447/commits/27a806cc44ddb4e8bc1551c37b995464f6569046. This was intended in this proposal as well, perhaps I should clarify that. The products are listed inline only in `GET /orders/{order_id}`.

I would suggest we add `/items` and linking them as optional, and allow listing them inline by default.

Reply:
But isn't listing a list of products for a specific collection basically the same as searching all products for a given collection ID? The same applies for the Item ID, I'd think.

Reply:
While I agree with @m-mohr that we should aim to reuse the existing STAC interfaces as much as possible, our implementation will inevitably have to take into account some limitations, specifically:

When listing `orderable` items:
- listing collection items for non-ordered products is only possible using Sentinel Hub's [3rd party data searching request](https://docs.sentinel-hub.com/api/latest/reference/#operation/dataImport_searchData);
- the generic part of the simple search API is _very_ limited (the only query parameters one can use in a generic way are `bounds`, `timeRange` and `maxCloudCoverage`);
- there are more options in the provider-specific "native search", but even these **don't currently allow fetching products by their ID** (except maybe with Planet data, where the native search is the most sophisticated);
- you always get all item data inline; there's paging to make 100+ products use-case manageable.

When listing `ordered` items:
- we can list all `product IDs` that were used as input when creating [a specific order](https://docs.sentinel-hub.com/api/latest/reference/#operation/dataImport_getOrder)
- after delivery, the delivery-related [endpoints](https://docs.sentinel-hub.com/api/latest/reference/#operation/dataImport_getOrderDelivery) again list only files and product IDs, but they don't contain STAC item metadata for the delivered products
- STAC item metadata is only available when the images are ingested into user collections (BYOC) within Sentinel Hub, and accessible via the specific collection's [catalog API](https://docs.sentinel-hub.com/api/latest/data/byoc/#catalog-api-capabilities)

My hope is that we can come up with a commercial data extension that is both reasonable as a standard interface as well as not-too-difficult to implement with the above constraints.

Reply:
> the generic part of the simple search API is very limited (the only query parameters one can use in a generic way are bounds, timeRange and maxCloudCoverage);

We already have `/search` in this spec, so I think the searching functionality of `/items` isn't really needed, and I also don't see a big utility in listing all available products.

IMO the only really useful thing is to have information about a single item on a separate endpoint, but as you mentioned this is exactly what we don't support atm. That's why I proposed to have it optional.

What we could also to is to have e.g. `GET /orders/{order_id}/products` or `GET /orders/{order_id}/items` (with pagination) and list all the products  metadata there.

Reply:
Given the limitations, I have two ideas:
1. Treat orders as new collections, i.e. each order creates a new collection at /collections and has then additional metadata for this specific order. You can then load specific orders via load_collection etc, but may end up with a lot of collections. You could then access the new data in two ways:
    1. via the original collection where the data for all orders can be accessed via load_collection.
    2. via the new collection where only data for the specific order can be accessed via load_collection
2. Do the same as above, but keep the new collections separate at /orders. Still, they should be full collections so that we can re-use as much code as possible from the collection implementations in the clients and back-ends, they just have some additional order-related metadata included.

Whether to choose 1 or 2 is more a "strategic" question on whether we want new orders to pop up in the collections endpoint or not.

Generally (i.e. for 1 and 2), no product metadata is inlined and you access them via .../items. Back-ends are free to choose whether they can make that only available for the new collections or also the old ones. You can give some information about products in that order, e.g. by listing product ids in collection summaries or so.

I think that should work having the limitations in mind, but maybe I missed something.

Reply:
> You could then access the new data in two ways:
> i. via the original collection where the data for all orders can be accessed via load_collection.
> ii. via the new collection where only data for the specific order can be accessed via load_collection

Doing something like this, i.e. saving purchased data to separate collections but allow loading them as one, is possible for us, as we have a "data fusion" feature, but if I'm not mistaken, requests to i.) would become linearly more expensive as new orders are added, so I don't think it's preferable (and it's definitely more complex). 

IMO it would be preferably to:

1. Have a single orderable "general" collection that contains data for all the orders and is used in `load_collection`, e.g. `collections/PLEIADES`
2. (Orderable) collections have `/items` where all ordered items are listed, e.g. `collections/PLEIADES/items`
3. I guess we will have to save the product Items on our backend and implement `collections/PLEIADES/items/{item-id}`, but only for the products used in orders, and not generally.
4. `orders/{order-id}` then use links to items

> I think that should work having the limitations in mind, but maybe I missed something. Happy to jump on another quick call to discuss details.

I would be happy to, when do you have time?

Reply:
Today 13:00, 16:00, tomorrow 11:00, 13:00, > 15:00, Fri 13:00

Reply:
Today at 16:00 is fine for me, I think you should have received the invite.

Reply:
Sorry about the meeting time limit, it's maximum by Webex on my account.

> Retrieve order information: GET /orders/:id -> contains metadata about the order (source and target collection id [might be the same or different], order extension fields, optionally links to product metadata), no inline product metadata

+ array of ordered `product-id`s

> Optional: Retrieve product metadata: GET /collections/:id/items and GET /collections/:id/items/:item_id -> may return error if not ingested yet, otherwise list of items. The items should contain links to the respective orders that made them available.

Ok let's go with this, as it makes the most sense.
My problem is just the Sentinel Hub implementation, as I don't think it is possible for items in `GET /collections/:id/items` to equal the ones the user queried with `/search` and ordered, with existing Sentinel Hub services (`GET /collections/:id/items` would contain only basic information about spatial and temporal extent).

However, this solves the issue of linking orders to items very nicely.

Reply:
> Can orders (in the future?) be created for multiple collections? e.g. ordering PLANET and MAXAR data at once?

I don't see this as a high priority requirement - cross-provider search might be useful sometimes, but cross-provider ordering seems too complicated to implement for not-so-much gain.

The interface would not need to change much, though - the `/order` endpoint is top-level, so we'd just need to include `collection_id` per each item and not have just one separate collection_id parameter. And we'd have to support multiple-output-collections-per-order. But I still suggest to leave this for some future upgrade.

Reply:
The question about cross-provider orders was actually related to the question whether /orders or /collections/:id/orders would be preferrable. And I now used /orders above to have exactly this flexibility, otherwise we could also go with /collections/:id/orders, I think

Reply:
> ... we could also go with /collections/:id/orders, I think

Even then, I think it's better to have `/orders`, so that all the user's orders are available in one place via  `GET /orders`. If one would want just a subset, we could at some point add filtering to this endpoint.

Reply:
I updated the README.md and I added the openAPI file for `/orders` and `/orders/{order_id}` endpoints. Should I add `/search`, `/items`, `queryables` etc. as well? Sadly it's not possible to reference them directly from Github because they contain relative references, so I'd have to copy most of it over.

Reply:
I'll have a look next week, thanks. I don't think we necessarily $ref the other endpoints as long as they are mentioned (and linked to) in the README.